engine/src/main/java/org/pentaho/di/core/auth/core/AuthenticationManager.java

```diff
@@ -0,0 +1,150 @@
+
+package org.pentaho.di.core.auth.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.pentaho.di.core.auth.core.impl.DefaultAuthenticationConsumerFactory;
+import org.pentaho.di.core.auth.core.impl.DefaultAuthenticationPerformerFactory;
+import org.pentaho.di.i18n.BaseMessages;
+
+public class AuthenticationManager {
+  private static final Class<?> PKG = AuthenticationManager.class;
+  private final Map<Class<?>, Map<Class<?>, Map<Class<?>, AuthenticationConsumerFactory<?, ?, ?>>>> factoryMap =
+      new HashMap<Class<?>, Map<Class<?>, Map<Class<?>, AuthenticationConsumerFactory<?, ?, ?>>>>();
+  private AuthenticationPerformerFactory authenticationPerformerFactory = new DefaultAuthenticationPerformerFactory();
+  private final List<AuthenticationProvider> authenticationProviders = new ArrayList<AuthenticationProvider>();
+
+  public void registerAuthenticationProvider( AuthenticationProvider authenticationProvider ) {
+    synchronized ( authenticationProviders ) {
+      authenticationProviders.add( authenticationProvider );
+    }
+  }
+
+  public boolean unregisterAuthenticationProvider( AuthenticationProvider authenticationProvider ) {
+    synchronized ( authenticationProviders ) {
+      return authenticationProviders.remove( authenticationProvider );
+    }
+  }
+
+  public <ReturnType, CreateArgType, ConsumedType> void registerConsumerFactory(
+      AuthenticationConsumerFactory<ReturnType, CreateArgType, ConsumedType> factory ) throws AuthenticationFactoryException {
+    if ( !factory.getConsumedType().isInterface()
+        && !AuthenticationProvider.class.isAssignableFrom( factory.getConsumedType() ) ) {
+      throw new AuthenticationFactoryException( BaseMessages.getString( PKG, "AuthenticationManager.ConsumedTypeError",
+          factory ) );
+    }
+
+    Map<Class<?>, AuthenticationConsumerFactory<?, ?, ?>> createTypeMap =
+        getRelevantConsumerFactoryMap( factory.getReturnType(), factory.getCreateArgType() );
+    synchronized ( createTypeMap ) {
+      createTypeMap.put( factory.getConsumedType(), factory );
+    }
+  }
+
+  public <ReturnType, ConsumedType> void registerConsumerClass(
+      Class<? extends AuthenticationConsumer<? extends ReturnType, ? extends ConsumedType>> consumerClass ) throws AuthenticationFactoryException {
+    registerConsumerFactory( new DefaultAuthenticationConsumerFactory( consumerClass ) );
+  }
+
+  public <ReturnType, CreateArgType, ConsumedType> List<AuthenticationPerformer<ReturnType, CreateArgType>>
+    getSupportedAuthenticationPerformers( Class<ReturnType> returnType, Class<CreateArgType> createArgType ) {
+    Map<Class<?>, AuthenticationConsumerFactory<?, ?, ?>> createTypeMap =
+        getRelevantConsumerFactoryMap( returnType, createArgType );
+    synchronized ( createTypeMap ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object createTypeMap synchronized on line 80 should not be re-assigned in the same synchronized block on line 81. After this re-assignment, you lose the reference to the synchronized object.